src/main/java/hudson/matrix/MatrixProject.java

```diff
@@ -199,14 +201,35 @@
      */
     private String childCustomWorkspace;
 
+    /**
+     * Lock to prevent project changes on different build at the same time
+     */
+    private ReadWriteLock buildLock;
```

Comment:
OK and because the function is not already present I simple add:
```
protected Object readResolve() {
    if (buildLock == null) {
        buildLock = new ReentrantReadWriteLock();
    }
    return this;
}
```
(as described [here](https://wiki.jenkins-ci.org/display/JENKINS/Hint+on+retaining+backward+compatibility))